File: impl/ocaml/src/import.ml
(* @import and @reference (SPEC §6.2). *)

open Ast
open Utils

let max_import_depth = 100

type params = { uri : string; layer : string option; media : string option; supports : string option }

(* Parses @import params; None for imports that must be left untouched. *)
let parse_import_params text =
  let rec drop_seps = function Value.Sep _ :: rest -> drop_seps rest | l -> l in
  match drop_seps (Value.parse text) with
  | Value.Word first :: rest -> (
      match unquote first.w with
      | None -> None
      | Some uri when has_prefix uri "data:" || has_prefix uri "http://" || has_prefix uri "https://" -> None
      | Some uri ->
          let layer = ref None and supports = ref None and media = ref [] and saw_media = ref false in
          List.iter
            (function
              | Value.Sep s -> if !media <> [] then media := !media @ [ s ]
              | Value.Fn f when f.fname = "layer" ->
                  if !supports <> None || !saw_media then
                    Twill_error.failf "`layer(...)` must appear before `supports(...)` and media queries in `@import %s`" text;
                  if !layer <> None then Twill_error.failf "Duplicate `layer(...)` in `@import %s`" text;
                  layer := Some (String.trim (Value.to_css f.fnodes))
              | Value.Fn f when f.fname = "supports" ->
                  if !saw_media then Twill_error.failf "`supports(...)` must appear before media queries in `@import %s`" text;
                  supports := Some (String.trim (Value.to_css f.fnodes))
              | Value.Fn f ->
                  saw_media := true;
                  media := !media @ [ Value.to_css [ Value.Fn f ] ]
              | Value.Word w when w.w = "layer" ->
                  if !supports <> None || !saw_media then
                    Twill_error.failf "`layer` must appear before `supports(...)` and media queries in `@import %s`" text;
                  layer := Some ""
              | Value.Word w ->
                  saw_media := true;
                  media := !media @ [ w.w ])
            rest;
          let media_text = String.trim (String.concat "" !media) in
          Some { uri; layer = !layer; supports = !supports; media = (if media_text = "" then None else Some media_text) })
  | _ -> None

(* Wraps imported nodes per SPEC §6.2. *)
let build_import_nodes nodes layer media supports =
  let root = match layer with Some l -> [ at_rule "@layer" l ~nodes ] | None -> nodes in
  let root = match media with Some m -> [ at_rule "@media" m ~nodes:root ] | None -> root in
  match supports with
  | Some s ->
      let condition = if has_prefix s "(" then s else "(" ^ s ^ ")" in
      [ at_rule "@supports" condition ~nodes:root ]
  | None -> root

(* Expands @import and @reference in place; returns the feature flags. *)
let rec substitute_at_imports_depth (ast : nodes) base (load : Builtin.loader) depth =
  let features = ref 0 in
  ignore
    (walk ast (fun n u ->
         match n with
         | At_rule at when at.name = "@import" || at.name = "@reference" -> (
             match parse_import_params at.params with
             | None -> Continue
             | Some parsed ->
                 let parsed = if at.name = "@reference" then { parsed with media = Some "reference" } else parsed in
                 if depth > max_import_depth then
                   Twill_error.failf "Exceeded maximum recursion depth while resolving `%s %s`" at.name at.params;
                 features := !features lor Features.at_import;
                 let loaded = load parsed.uri base in
                 let imported = ref (Parser.parse loaded.content) in
                 features := !features lor substitute_at_imports_depth imported loaded.base load (depth + 1);
                 let wrapped =
                   build_import_nodes
                     [ context (ctx_of_list [ ("base", loaded.base) ]) ~nodes:!imported ]
                     parsed.layer parsed.media parsed.supports
                 in
                 replace_with u [ context ctx_empty ~nodes:wrapped ];
                 Skip)
         | _ -> Continue));
  !features

let substitute_at_imports ast base load = substitute_at_imports_depth ast base load 0

File: impl/go/import.go
package twill

import "strings"

const maxImportDepth = 100

// ImportParams are the parsed params of an @import rule.
type ImportParams struct {
	URI      string
	Layer    *string
	Media    *string
	Supports *string
}

func strPtr(s string) *string { return &s }

// ParseImportParams parses @import params. It returns nil for imports that
// must be left untouched (url(), data:, http(s)).
func ParseImportParams(params string) (*ImportParams, error) {
	nodes := ParseValue(params)
	for len(nodes) > 0 {
		if _, ok := nodes[0].(*ValueSeparator); ok {
			nodes = nodes[1:]
			continue
		}
		break
	}
	if len(nodes) == 0 {
		return nil, nil
	}
	first, ok := nodes[0].(*ValueWord)
	if !ok {
		return nil, nil
	}
	uri, ok := Unquote(first.Value)
	if !ok {
		return nil, nil
	}
	if strings.HasPrefix(uri, "data:") || strings.HasPrefix(uri, "http://") || strings.HasPrefix(uri, "https://") {
		return nil, nil
	}
	nodes = nodes[1:]

	result := &ImportParams{URI: uri}
	var media []string
	sawMedia := false
	for _, n := range nodes {
		switch n := n.(type) {
		case *ValueSeparator:
			if len(media) > 0 {
				media = append(media, n.Value)
			}
		case *ValueFunction:
			if n.Name == "layer" {
				if result.Supports != nil || sawMedia {
					return nil, Errorf("`layer(...)` must appear before `supports(...)` and media queries in `@import %s`", params)
				}
				if result.Layer != nil {
					return nil, Errorf("Duplicate `layer(...)` in `@import %s`", params)
				}
				result.Layer = strPtr(strings.TrimSpace(ValueToCSS(n.Nodes)))
				continue
			}
			if n.Name == "supports" {
				if sawMedia {
					return nil, Errorf("`supports(...)` must appear before media queries in `@import %s`", params)
				}
				result.Supports = strPtr(strings.TrimSpace(ValueToCSS(n.Nodes)))
				continue
			}
			sawMedia = true
			media = append(media, ValueToCSS([]ValueNode{n}))
		case *ValueWord:
			if n.Value == "layer" {
				if result.Supports != nil || sawMedia {
					return nil, Errorf("`layer` must appear before `supports(...)` and media queries in `@import %s`", params)
				}
				result.Layer = strPtr("")
				continue
			}
			sawMedia = true
			media = append(media, n.Value)
		}
	}
	if text := strings.TrimSpace(strings.Join(media, "")); text != "" {
		result.Media = strPtr(text)
	}
	return result, nil
}

// BuildImportNodes wraps imported nodes per SPEC §6.2.
func BuildImportNodes(nodes []Node, layer, media, supports *string) []Node {
	root := nodes
	if layer != nil {
		root = []Node{NewAtRule("@layer", *layer, root...)}
	}
	if media != nil {
		root = []Node{NewAtRule("@media", *media, root...)}
	}
	if supports != nil {
		condition := *supports
		if !strings.HasPrefix(condition, "(") {
			condition = "(" + condition + ")"
		}
		root = []Node{NewAtRule("@supports", condition, root...)}
	}
	return root
}

// SubstituteAtImports expands @import and @reference in place. It returns
// FeatureAtImport whenever an import was resolved.
func SubstituteAtImports(ast *[]Node, base string, load StylesheetLoader) (Features, error) {
	return substituteAtImports(ast, base, load, 0)
}

func substituteAtImports(ast *[]Node, base string, load StylesheetLoader, depth int) (Features, error) {
	var features Features
	var failure error
	Walk(ast, func(n Node, u *WalkUtils) WalkAction {
		at, ok := n.(*AtRule)
		if !ok || (at.Name != "@import" && at.Name != "@reference") {
			return Continue
		}
		parsed, err := ParseImportParams(at.Params)
		if err != nil {
			failure = err
			return Stop
		}
		if parsed == nil {
			return Continue
		}
		if at.Name == "@reference" {
			parsed.Media = strPtr("reference")
		}
		if depth > maxImportDepth {
			failure = Errorf("Exceeded maximum recursion depth while resolving `%s %s`", at.Name, at.Params)
			return Stop
		}
		features |= FeatureAtImport
		loaded, err := load(parsed.URI, base)
		if err != nil {
			failure = err
			return Stop
		}
		imported, err := Parse(loaded.Content)
		if err != nil {
			failure = err
			return Stop
		}
		nested, err := substituteAtImports(&imported, loaded.Base, load, depth+1)
		if err != nil {
			failure = err
			return Stop
		}
		features |= nested
		wrapped := BuildImportNodes([]Node{NewContext(ContextMap{"base": loaded.Base}, imported...)}, parsed.Layer, parsed.Media, parsed.Supports)
		u.ReplaceWith(NewContext(ContextMap{}, wrapped...))
		return Skip
	})
	if failure != nil {
		return 0, failure
	}
	return features, nil
}
